pkg/lib/config/authenticator_test.go

```diff
@@ -0,0 +1,220 @@
+package config_test
+
+import (
+	"os"
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+)
+
+func TestValidPeriodDeprecateFlow(t *testing.T) {
+	Convey("Email Code Mode", t, func() {
+		test := func(inputYAML string, mode string) {
+			cfg, err := config.Parse([]byte(inputYAML))
+			So(err, ShouldBeNil)
+
+      var data []byte
+      switch mode {
+      case "code":
+        data, err = os.ReadFile("testdata/code_validperiod_test.yaml")
+        if err != nil {
+          panic(err)
+        }
+      case "login_link":
+        data, err = os.ReadFile("testdata/link_validperiod_test.yaml")
+        if err != nil {
+          panic(err)
+        }
+      default:
+        data, err = os.ReadFile("testdata/default_validperiod_test.yaml")
+        if err != nil {
+          panic(err)
+        }
+      }
+
+			defaultConfig, err := config.Parse(data)
+			So(err, ShouldBeNil)
+
+			So(cfg, ShouldResemble, defaultConfig)
+		}
+
+		// Both deprecated with email_otp_mode: code
+		test(`
+id: test
+http:
+  public_origin: http://test
+authenticator:
+  oob_otp:
+    email:
+      email_otp_mode: code
+      maximum: 99
+      code_valid_period: 321s
+    sms:
+      maximum: 99
+      phone_otp_mode: whatsapp_sms
+      code_valid_period: 321s
+verification:
+  code_valid_period: 321s
+`, "code")
```

Comment:
Instead of writing extra code to map input to output. Create a new directory and place a input file and a output file per case.

```
testdata/
  authenticator_code_valid_period_deprecation/
    case-1-read_deprecated_config_code.input.yaml
    case-1-read_deprecated_config_code.output.yaml
```

```
func TestAuthenticatorCodeValidPeriodDeprecation(t *testing.T) {
  Convey("authenticator_code_valid_period_deprecation", t, func() {
    entries := os.ReadDir("testdata/authenticator_code_valid_period_deprecation")
    for _, entry := range entries {
      withoutYAML := Basename(entry.Name)
      withoutInputOutput := BaseName(withoutYAML)
      Convey(withoutInputOutput, func() {
        inputPath := Join(
        inputBytes := os.ReadFile("testdata/authenticator_code_valid_period_deprecation", withoutInputOutput, "input", "yaml")
        outBytes := ...
      })
    }
  })
}
```